mathics/core/formatter.py

```diff
@@ -119,3 +151,386 @@ def add_conversion_fn(cls, module_fn_name=None) -> None:
+            do_format = element_formatters.get(type(formatted), do_format_element)
+            result = do_format(formatted, evaluation, form)
+            if include_form:
+                result = Expression(form, result)
+            return result
+
+        # If the expression is still enclosed by a Format,
+        # iterate.
+        # If the expression is not atomic or of certain
+        # specific cases, iterate over the elements.
+        head = expr.get_head()
+        if head in formats:
+            do_format = element_formatters.get(type(element), do_format_element)
+            expr = do_format(expr, evaluation, form)
+        elif (
+            head is not SymbolNumberForm
+            and not isinstance(expr, (Atom, BoxElement))
+            and head not in (SymbolGraphics, SymbolGraphics3D)
+        ):
+            # print("Not inside graphics or numberform, and not is atom")
+            new_elements = [
+                element_formatters.get(type(element), do_format_element)(
+                    element, evaluation, form
+                )
+                for element in expr.elements
+            ]
+            expr_head = expr.head
+            do_format = element_formatters.get(type(expr_head), do_format_element)
+            expr = Expression(do_format(expr_head, evaluation, form), *new_elements)
+        if include_form:
+            expr = Expression(form, expr)
+        return expr
+    finally:
+        evaluation.dec_recursion_depth()
+
+
+def do_format_rational(
+    element: BaseElement, evaluation: Evaluation, form: Symbol
+) -> BaseElement:
+    if form is SymbolFullForm:
+        return do_format_expression(
+            Expression(
+                Expression(SymbolHoldForm, SymbolRational),
+                element.numerator(),
+                element.denominator(),
+            ),
+            evaluation,
+            form,
+        )
+    else:
+        numerator = element.numerator()
+        minus = numerator.value < 0
+        if minus:
+            numerator = Integer(-numerator.value)
+        result = Expression(SymbolDivide, numerator, element.denominator())
+        if minus:
+            result = Expression(SymbolMinus, result)
+        result = Expression(SymbolHoldForm, result)
+        return do_format_expression(result, evaluation, form)
+
+
+def do_format_complex(
+    element: BaseElement, evaluation: Evaluation, form: Symbol
+) -> BaseElement:
+    if form is SymbolFullForm:
+        return do_format_expression(
+            Expression(
+                Expression(SymbolHoldForm, SymbolComplex), element.real, element.imag
+            ),
+            evaluation,
+            form,
+        )
+
+    parts: typing.List[Any] = []
+    if element.is_machine_precision() or not element.real.is_zero:
+        parts.append(element.real)
+    if element.imag.sameQ(Integer(1)):
+        parts.append(SymbolI)
+    else:
+        parts.append(Expression(SymbolTimes, element.imag, SymbolI))
```

Comment:
Just an idle curiosity, wonder if we can use that fancy infix notation stuff like `element.imag * SymbollI`.  I have noticed other places that are kind of like this too.  It is just something I noticed and wondered about. Of course, this isn't wrong.

Reply:
This way of doing it (first building a `Times` expression, and then asking for formatting it, instead of building an Infix notation) gives some flexibility in defining different ways to represent a product in different formats. For example, the expression `Times[a,I]`, in `InputForm` is represented as  `a*I`, but in `StandardForm` is `a I`. This behavior can also be modified by the user inside a session, so I guess that this is the right way to implement it.

Reply:
I am not sure I understand because the infix notation is supposed to do the same thing: it is just syntactic sugar. 

But no matter - never was a big deal either way.